### PR TITLE
admin: Richer HTML home page with forms for params

### PR DESCRIPTION
**Pull request by jmarantz**
Commit Message:
High level: creates an HTML UI to tweak query params for admin endpoints.

This PR covers a few things and could be broken up.

 * adds ParamDescriptor abstraction and gives every endpoint a vector of those so the query-params can be specified before the admin page is visited.

 * Re-organizes CSS and HTML fragments so they are in their own files. This makes it easy to iterate on them as editors can use the right syntax-highlighting each file type, and you can hack the generator script during development to reference the file from a local server so you can iterate on the web markup without recompiling/relinking C++ or even restarting the server.

 * Enable adjusting query-params for admin points from the admin home page:

admin_screenshot

 * Adds an "html" format option for /stats and allow incremental adjustment of settings from that view also.

admin_stats_screenshot

 * Adds selection of types (text-readouts, gauges, counters, histograms, or All) within /stats.

Some of these changes could be broken out as their own PRs if this is too large to review.

Additional Description: 
This is was broken out from #18670 which will enable a hierarchical view of the stat names, made much more efficient by employing Stats::Scope to bound the traversal that must be done by the server.
Risk Level: medium -- this adds a novel new mechanism to build HTML and CSS files into the C++ binary as string constants, to simplify deployment. This appears to pass CI on non-Unix platforms so I think this is correct.
Tests: //test/server/..., manually clicking around admin page, and validating HTML with https://validator.w3.org/nu/#textarea
Docs Changes: will be needed in this PR but would like to get traction on code before adding that.
Platform Specific Features: n/a

### CODE REVIEW

**Comment by jmarantz** [comment]
@dmitri-d let me know if you'd prefer to break this up prior to review.

**Comment by jmarantz** [comment]
We are rolling back code on which this depends, so converting to draft mode for now. See https://github.com/envoyproxy/envoy/pull/20835

**Comment by jmarantz** [comment]
Re-opening as the related bug has now been resolved.

I'm OK with breaking this up if preferred e.g. based on the bullet points in the PR description.

**Comment by jmarantz** [comment]
Matt & Greg -- it'd be great to get your take on this at a high level at least; happy to find a first pass detailed review elsewhere.

**Comment by mattklein123** [comment]
High level extremely quick comment: What do you think about somehow moving this out to an extension or making it compile time switchable? I don't think this will be widely used and it would especially be nice to be able to compile something like this out for Envoy Mobile? WDYT?

**Comment by jmarantz** [comment]
This makes sense and I agree that it'd be great to have a minimal E-M build. What is the right partitioning?

Will there be an admin endpoint? Stats sinks? E.g. currently the admin "/" provides a minimal HTML response, whereas I think it used to, a long time back, respond with a text/plain list of endpoints. Is that a better behavior for E-M? Are things like Prometheus used in E-M?

 I have to admit I don't have a great intuition about how this is used in a mobile setting; I understand why a proxy is good but idk how to monitor it.

**Comment by mattklein123** [comment]
> Will there be an admin endpoint? Stats sinks? E.g. currently the admin "/" provides a minimal HTML response, whereas I think it used to, a long time back, respond with a text/plain list of endpoints. Is that a better behavior for E-M? Are things like Prometheus used in E-M?

I think in general very few people use any kind of HTML/browser interaction with the admin endpoint when running at scale. IMO this is really great for local examples and learning Envoy and maybe targeted debugging but in general seems like something that wouldn't be widely used, even on server.

Could we potentially just have all the HTML stuff be behind a compile time option even if not an extension? If we want to do an extension, another way of thinking about it would be as a bootstrap extension that could register different admin handlers (HTML variants)?

**Comment by ggreenway** [comment]
I agree that I don't think it would be widely used, and I like the idea of adding it in via an extension. I don't think I've ever accessed the admin endpoint with anything other than curl.

We've discussed admin listener extensions a few times. Would it make sense to support an admin listener filter chain? Another recent request I saw was for compressing stats responses, which is another good use case for adding filters.

I think previous brainstorming on this topic was to make the admin endpoint just a regular filter to put on a normal listener. This was difficult because many admin actions need to happen on the main thread, but listeners handle traffic on worker threads. What if instead we kept the admin listener as special and always on the main thread, but still allow some configurable filters?

But it's also possible that the admin listener is just too different from normal dataplane listeners, and this isn't worth the effort it would take.

**Comment by jmarantz** [comment]
I don't think it's a problem to have the admin filter run on a worker thread. It can just post to the main thread if it needs to read data that way. If that enables compression to be added into the chain, and to make it easier for admin endpoints to be extensions that SGTM.

WDYT?

**Comment by ggreenway** [comment]
I'm not opposed to that. There has been previous work and discussion in this area. See #11367 and #2763. But I'd love to solve this "for real" since it has come up repeatedly over the years.

**Comment by jmarantz** [comment]
One other related thought -- if no one other than us uses the html version of the admin endpoint should we consider removing it from Envoy OSS and put back the text/plain list-of-endpoints that was originally there? I think we can add the alternate entry-point here without much difficulty.

That might also have the benefit of reducing the current code footprint and making E-M smaller.

**Comment by jmarantz** [comment]
Also for this PR I think what I'll do first is fork off the Params refactor since is a straight code simplification and will make it easier to than add RE2-based filtering, as std::regex is a bottleneck.

**Comment by mattklein123** [comment]
Yeah I would be in favor of removing the existing HTML and making it all optional somehow.

**Comment by mattklein123** [comment]
Unassigned myself. Please reassign when not a draft.

**Comment by jmarantz** [comment]
OK I'm going to open this up for review. I think it functionally does what I want and I'd like to start a dialog about a couple of the novel things in this PR.

**Comment by jmarantz** [comment]
Re-assigning to Matt now that this is not a draft. I think this passes CI also though it seems to be stuck in some steps now. I'm sure there will be suggestions anyway :)

File: source/server/admin/stats_html_render.cc
Comment (Mythra) [reply source/server/admin/stats_html_render.cc:127]: (Note this will also allow us to remove things like "pointer: cursor" and things that shouldn't apply to links in terms of styling)

**Review by Mythra — COMMENTED** [review]
From the perspective of solely just the HTML Generated (not talking about the actual C++ yet since I was asked to just review the HTML first) -- I think this is a good start. I also took a quick peak at this with my screen-reader. Only found two things I think make sense to change for an initial bare-bones admin page widely distributed.

If anything isn't clear, or you'd like help testing/understanding anything please don't hesitate to reach out. Once those fixes are up I can do a quick pass over the full thing including the C++ code generating it!

Thanks so much for pushing this forward.

File: source/server/admin/stats_html_render.cc
Comment (jmarantz) [reply source/server/admin/stats_html_render.cc:127]: The reason I think it needs to be a form is that it needs to collect some input parameters. This page is a little more complex as it shows distinct form parameters for each endpoint. The endpoints that mutate state are rendered as buttons and issue POSTs. The endpoints that just return info are rendered as links and issue GETs but they both need to collect their associated parameters.

In the screen-shot in the description you can see a POST example "drain_listeners" and you can see a GET example "config_dump".

I can also attach the generated HTML for this if that helps.

File: source/server/admin/stats_html_render.cc
Comment (jmarantz) [reply source/server/admin/stats_html_render.cc:136]: OK let me take a look at that. Thanks for catching this!

File: source/server/admin/stats_html_render.cc
Comment (jmarantz) [reply source/server/admin/stats_html_render.cc:127]: Oh I think I see your point -- many of the endpoints do not have parameters and they could be simple links. I'd be OK with that if that makes the screen-reader experience better.

What's the easiest way for me to see that that looks like with a screen reader?

**Comment by jmarantz** [comment]
[admin.zip](https://github.com/envoyproxy/envoy/files/9123430/admin.zip)

File: source/server/admin/stats_html_render.cc
Comment (jmarantz) [reply source/server/admin/stats_html_render.cc:136]: I attached the admin homepage html as a zip file to this PR

File: source/server/admin/stats_html_render.cc
Comment (Mythra) [reply source/server/admin/stats_html_render.cc:127]: Yeah I'm purely talking about the ones without parameters here (the ones with it should ideally be so too, but that requires JS which is a much longer discussion I don't think we wanna take on, plus having the page work without JS is nice). Sorry if that wasn't clear!

As for testing with a screenreader yourself it depends a bit on the OS, if you're using a Mac it's just in settings as voiceover, and they should also have a tutorial. On windows there isn't a good one out of the box , you'd want to use NVDA (since Jaws the other popular one isn't free), though there's no tutorial. Linux has ORCA, but that's requires lots of finagling in my experience. 

Screenreader testing is also usually very context dependent as it can be hard to know how something should be "read". I'd be happy to discuss this in slack, or over a call, but the core things this helps with are:

1. Most screenreaders have a concept of a "rotor" which is for quickly navigating landmarks/forms/etc. of a page, the more helpful we can make this rotor by say not putting a bunch of forms on, the easier it is to quickly navigate.
2. On Windows, screenreaders have separate modes depending on if you're in a "form" or not. Many users switch to form mode whenever they hear they're entering a fork to better handle it. To get presented a form with no content is quite confusing when you can't see there's no data to input.

Comment (jmarantz) [reply source/server/admin/stats_html_render.cc:132]: Thanks -- I think I did the corrections you suggested, though the exact structure you specified was not correct according an HTML validator site. I corrected that and tried the Mac screen reader, though I couldn't detect any difference in the behavior; it seemed to work OK.

In the form that's committed to the PR now, I get only one validation error for the inline favicon, though I think the validator is wrong in that case; the favicon works in Firefox. It doesn't show up in Chrome for me because Chrome overrides the favicon for HTTP sites as a warning to users that the connection is not private.

PTAL.

**Comment by jmarantz** [comment]
sorry hold off on this a moment -- getting the associations right between param elements and forms has broken the form submission data :)

**Comment by jmarantz** [comment]
OK this seems to work. An example of a fragment with one parameter is here:
```
<tr class='vert-space'><td></td><td></td></tr>
<tr>
  <td class='home-data'><form action='init_dump' method='get' id='init_dump' class='home-form'>
    <button class='button-as-link'>init_dump</button>
  </form></td>
  <td class='home-data'>dump current Envoy init manager information (experimental)</td>
</tr>
<tr>
  <td class='option'><input type='text' name='mask' id='param-12-init_dump-mask' form='init_dump' /></td>
  <td class='home-data'><label for='param-12-init_dump-mask'>The desired component to dump unready targets. The mask is parsed as a ProtobufWkt::FieldMask. For example, get the unready targets of all listeners with /init_dump?mask=listener`</label></td>
</tr>
```

File: source/server/admin/stats_html_render.cc
Comment (jmarantz) [reply source/server/admin/stats_html_render.cc]: I have updated the GET endpoints without parameters as links. I  am not 100% sure this change makes things better.

If it does, should we also render the POST endpoints without parameters without a form? Maybe a simple button?

Comment (Mythra) [reply source/server/admin/stats_html_render.cc]: Hey thanks for doing this! I can take a look here, I imagine it'll improve the rotor for OSX, and the Windows "Form" vs "Non-Form" mode. I'm happy to take a peek with both.

As for "POST" parameters, we'd either need to do a form or something with javascript to send a post on button click. I don't think we need to complicate it at this stage by adding JS to the site, should we make it in the long term? Maybe. I wouldn't worry about that for now.

**Review by Mythra — APPROVED** [review]
This feels like a great start to me.

**Comment by jmarantz** [comment]
@mattklein123 does it make sense for you to take a look now or do you want me to get another maintainer to look over the C++ and BUILD aspects of this first?

**Comment by jmarantz** [comment]
thanks @Mythra !

File: source/server/admin/stats_params.h
Comment (pradeepcrao) [inline source/server/admin/stats_params.h:17]: Should the namespace be less generically named (since the enclosing namespaces aren't specifically stat related)?
(I don't feel too strongly about this)

File: source/server/admin/stats_params.cc
Comment (pradeepcrao) [inline source/server/admin/stats_params.cc:58]: maybe including the param value here will be helpful?

File: source/server/admin/stats_params.h
Comment (pradeepcrao) [inline source/server/admin/stats_params.h:58]: Where is this used?

**Review by pradeepcrao — COMMENTED** [review]
Thank you so much for doing this!

I only have a few nits after a first pass. Doing a more detailed review now.

File: source/server/admin/stats_params.h
Comment (jmarantz) [reply source/server/admin/stats_params.h]: Done

File: source/server/admin/stats_params.h
Comment (jmarantz) [reply source/server/admin/stats_params.h]: good catch! I think it is used in the follow-up PR but will add it there. Removed.

File: source/server/admin/stats_html_render.cc
Comment (jmarantz) [reply source/server/admin/stats_html_render.cc]: agreed --  I think at one point this was used twice but not any longer. Inlined.

File: source/server/admin/stats_html_render.cc
Comment (jmarantz) [reply source/server/admin/stats_html_render.cc]: probably not, but to be consistent I'll do that.

File: source/server/admin/stats_params.cc
Comment (jmarantz) [reply source/server/admin/stats_params.cc:58]: In general I've learned to avoid including bogus input in an error response. It is an invitation to have an XSS.

Of course I can sanitize it but it's easier to just not include it.

Comment (pradeepcrao) [reply source/server/admin/stats_params.cc:58]: Sounds good.

File: test/server/admin/admin_test.cc
Comment (jmarantz) [reply test/server/admin/admin_test.cc:168]: How about `"One of (cumulative,, disjoint, none)"`

Comment (pradeepcrao) [reply test/server/admin/admin_test.cc:168]: looks good (I am guessing you meant to put only 1 comma after `cumulative`)

File: source/server/admin/admin.cc
Comment (jmarantz) [reply source/server/admin/admin.cc:115]: I realized that and was thinking of doing filling in all the new parameters that were added since I first wrote this code as a follow up. However I'm glad you asked about this because this brings a new complexity. `/logging` is a mutating operation, and thus must be issued as a POST. POST params are transmitted from the browser to the server as POST params in the request body rather than query-params, and there's not currently a path to get get them into the admin handlers. I'm going to have to iterate on this a bit.

I'll get that sorted before asking for the next round of comments.

/wait

Comment (jmarantz) [reply source/server/admin/admin.cc:105]: I decided to back out my addition of POST params in this PR and will follow up. I got this working but it was too hacky. I need to make some infrastructural changes to get POST params to work well, and also to handling the "/logging" semantics for an optional enum param.

**Review by jmarantz — COMMENTED** [review]
OK I think this is ready for another round of review.

A follow-up will be needed to finish the POST params.

File: source/server/admin/admin.cc
Comment (jmarantz) [reply source/server/admin/admin.cc:102]: done.

File: test/server/admin/stats_handler_test.cc
Comment (pradeepcrao) [inline test/server/admin/stats_handler_test.cc:212]: Nice!

**Review by pradeepcrao — COMMENTED** [review]
LGTM

**Review by yanavlasov — APPROVED** [review]

**Pull request merged**
